pyomo/core/base/var.py

```diff
@@ -408,20 +385,22 @@ def set_value(self, val, skip_validation=False):
                     extra={'id':'W1002'},
                 )
 
+        if self._value == val and not self.stale:
+            # do not update the value / set the stale flag if we are not
+            # changing the value and the variable is currently not stale
+            return
```

Comment:
Is this desired? I would think the stale flag should always get updated when calling `set_value`, even if the the value is the same. What if the value just happens to be the same?

```
>>> StaleFlagManager.mark_all_as_stale(delayed=False)
>>> m.x.value = 0
>>> m.y.value = 1
>>> StaleFlagManager.mark_all_as_stale(delayed=True)
>>> m.x.stale, m.y.stale
(False, False)
>>> m.x.value = 0; m.y.value = 0
>>> m.x.stale, m.y.stale
(True, False)
```

I don't think this is correct???

Reply:
Desired is of course open for discussion, but it is intentional given this implementation.  The idea is that "stale" indicates that the variable value is potentially inconsistent with the results from the most recent "batch" model update (e.g., from running a solver or loading a solution).  The implicit assumption is that after a solve, a user / algorithm may update variables: if the variable is currently stale, the assumption is that the new value is to update the variable to be consistent with the solve result.  However, if the variable is not stale, then if the variable value is being changed, then it will (almost by definition) *not* be consistent with the previous solve, so the current stale flag gets advanced (marking all variables as stale) before assigning the new value.  However, if the variable value is not changing, then it is still consistent with the previous solve, and since it is still consistent, we don't have to advance the stale flag.

The more I think about it, I could be convinced that a user updating a variable value that is currently not stale is *intentionally* updating the model, so we could / should remove this case (and make it so that *any* update to a non-stale variable would trigger the delayed "mark as stale" action).  The workaround is of course for the user to mark the variable as stale before updating the value (something that we do as part of solution updates in MindtPy).

Reply:
At the Dev call we decided that the expected behavior for `stale` should follow @michaelbynum's original intuition.  The PR has been updated (including tests for the expected behavior).